File: packages/infermap-js/src/core/calibration.ts
// Confidence calibration — post-assignment transform of raw scores into
// calibrated probabilities. Mirrors infermap/calibration.py.
//
// A calibrator does NOT change which mappings the engine picks; it only
// relabels the confidence attached to each picked mapping. JSON round-trip
// uses a `kind` discriminator for portability between Python and JS.

export interface CalibratorJSON {
  kind: string;
  [k: string]: unknown;
}

export interface Calibrator {
  readonly kind: string;
  fit(scores: readonly number[], correct: readonly number[]): void;
  transform(scores: readonly number[]): number[];
  toJSON(): CalibratorJSON;
}

// ---------- Identity ----------

export class IdentityCalibrator implements Calibrator {
  readonly kind = "identity";
  fit(_scores: readonly number[], _correct: readonly number[]): void {
    // no-op
  }
  transform(scores: readonly number[]): number[] {
    return scores.map((s) => s);
  }
  toJSON(): CalibratorJSON {
    return { kind: this.kind };
  }
  static fromJSON(_obj: CalibratorJSON): IdentityCalibrator {
    return new IdentityCalibrator();
  }
}

// ---------- Isotonic (pool-adjacent-violators) ----------

export class IsotonicCalibrator implements Calibrator {
  readonly kind = "isotonic";
  x: number[];
  y: number[];

  constructor(x: number[] = [0, 1], y: number[] = [0, 1]) {
    this.x = x.slice();
    this.y = y.slice();
  }

  fit(scores: readonly number[], correct: readonly number[]): void {
    const n = scores.length;
    if (n === 0) return;
    // Stable sort by score ascending.
    const order = Array.from({ length: n }, (_, i) => i).sort((a, b) => {
      const d = scores[a]! - scores[b]!;
      return d !== 0 ? d : a - b;
    });
    const xs = order.map((i) => scores[i]!);
    const ys = order.map((i) => correct[i]!);

    // Pool-adjacent-violators.
    const vals: number[] = [];
    const wts: number[] = [];
    for (const y of ys) {
      vals.push(y);
      wts.push(1);
      while (vals.length >= 2 && vals[vals.length - 2]! > vals[vals.length - 1]!) {
        const v2 = vals.pop()!;
        const w2 = wts.pop()!;
        const v1 = vals.pop()!;
        const w1 = wts.pop()!;
        vals.push((v1 * w1 + v2 * w2) / (w1 + w2));
        wts.push(w1 + w2);
      }
    }

    // Expand pooled values back to per-point fitted ys.
    const fitted = new Array<number>(n);
    let idx = 0;
    for (let b = 0; b < vals.length; b++) {
      const k = Math.round(wts[b]!);
      for (let j = 0; j < k; j++) fitted[idx++] = vals[b]!;
    }

    // Dedup by xs: average fitted values for identical xs.
    const uniqX: number[] = [];
    const uniqY: number[] = [];
    const counts: number[] = [];
    for (let i = 0; i < n; i++) {
      const xv = xs[i]!;
      if (uniqX.length > 0 && uniqX[uniqX.length - 1] === xv) {
        uniqY[uniqY.length - 1]! += fitted[i]!;
        counts[counts.length - 1]! += 1;
      } else {
        uniqX.push(xv);
        uniqY.push(fitted[i]!);
        counts.push(1);
      }
    }
    for (let i = 0; i < uniqY.length; i++) uniqY[i] = uniqY[i]! / counts[i]!;

    // Enforce monotonicity (defensive) and clip to [0, 1].
    for (let i = 1; i < uniqY.length; i++) {
      if (uniqY[i]! < uniqY[i - 1]!) uniqY[i] = uniqY[i - 1]!;
    }
    for (let i = 0; i < uniqY.length; i++) {
      uniqY[i] = Math.max(0, Math.min(1, uniqY[i]!));
    }

    this.x = uniqX;
    this.y = uniqY;
  }

  transform(scores: readonly number[]): number[] {
    if (this.x.length === 0) return scores.map((s) => s);
    const out = new Array<number>(scores.length);
    const x = this.x;
    const y = this.y;
    const last = x.length - 1;
    for (let i = 0; i < scores.length; i++) {
      const s = scores[i]!;
      if (s <= x[0]!) {
        out[i] = y[0]!;
      } else if (s >= x[last]!) {
        out[i] = y[last]!;
      } else {
        // Binary search.
        let lo = 0;
        let hi = last;
        while (lo + 1 < hi) {
          const mid = (lo + hi) >> 1;
          if (x[mid]! <= s) lo = mid;
          else hi = mid;
        }
        const x0 = x[lo]!;
        const x1 = x[hi]!;
        const y0 = y[lo]!;
        const y1 = y[hi]!;
        out[i] = x1 === x0 ? y0 : y0 + ((s - x0) * (y1 - y0)) / (x1 - x0);
      }
    }
    return out;
  }

  toJSON(): CalibratorJSON {
    return { kind: this.kind, x: this.x.slice(), y: this.y.slice() };
  }

  static fromJSON(obj: CalibratorJSON): IsotonicCalibrator {
    const x = Array.isArray(obj["x"]) ? (obj["x"] as number[]).map(Number) : [0, 1];
    const y = Array.isArray(obj["y"]) ? (obj["y"] as number[]).map(Number) : [0, 1];
    return new IsotonicCalibrator(x, y);
  }
}

// ---------- Platt (sigmoid) ----------

function logSigmoid(z: number): number {
  // -log(1 + exp(-z)) numerically stable
  if (z >= 0) return -Math.log1p(Math.exp(-z));
  return z - Math.log1p(Math.exp(z));
}

export class PlattCalibrator implements Calibrator {
  readonly kind = "platt";
  a: number;
  b: number;

  constructor(a = 1.0, b = 0.0) {
    this.a = a;
    this.b = b;
  }

  fit(scores: readonly number[], correct: readonly number[]): void {
    if (scores.length === 0) return;
    // Minimize binary cross-entropy over (a, b) via Nelder-Mead.
    const nll = (params: [number, number]): number => {
      const [a, b] = params;
      let total = 0;
      for (let i = 0; i < scores.length; i++) {
        const z = a * scores[i]! + b;
        const c = correct[i]!;
        // c * log sigmoid(z) + (1-c) * log(1-sigmoid(z))
        // log(1-sigmoid(z)) = logSigmoid(-z)
        total += c * logSigmoid(z) + (1 - c) * logSigmoid(-z);
      }
      return -total;
    };
    const [a, b] = nelderMead2D(nll, [1.0, 0.0], 1e-6, 500);
    this.a = a;
    this.b = b;
  }

  transform(scores: readonly number[]): number[] {
    const out = new Array<number>(scores.length);
    for (let i = 0; i < scores.length; i++) {
      const z = this.a * scores[i]! + this.b;
      // Numerically stable sigmoid.
      out[i] = z >= 0 ? 1 / (1 + Math.exp(-z)) : Math.exp(z) / (1 + Math.exp(z));
    }
    return out;
  }

  toJSON(): CalibratorJSON {
    return { kind: this.kind, a: this.a, b: this.b };
  }

  static fromJSON(obj: CalibratorJSON): PlattCalibrator {
    return new PlattCalibrator(Number(obj["a"] ?? 1), Number(obj["b"] ?? 0));
  }
}

/**
 * Minimal Nelder-Mead for 2D functions. Used by Platt calibration because
 * scipy.optimize isn't available in JS. Converges to `tol` or `maxIter`,
 * whichever first.
 */
function nelderMead2D(
  f: (p: [number, number]) => number,
  x0: [number, number],
  tol: number,
  maxIter: number
): [number, number] {
  // Build an initial simplex.
  const step = 0.1;
  type P = [number, number];
  const simplex: P[] = [
    [x0[0], x0[1]],
    [x0[0] + step, x0[1]],
    [x0[0], x0[1] + step],
  ];
  let fs = simplex.map((p) => f(p));

  const alpha = 1;
  const gamma = 2;
  const rho = 0.5;
  const sigma = 0.5;

  for (let iter = 0; iter < maxIter; iter++) {
    // Sort by f ascending.
    const idx = [0, 1, 2].sort((a, b) => fs[a]! - fs[b]!);
    const s: P[] = [simplex[idx[0]!]!, simplex[idx[1]!]!, simplex[idx[2]!]!];
    const fsorted = [fs[idx[0]!]!, fs[idx[1]!]!, fs[idx[2]!]!];
    simplex[0] = s[0]!;
    simplex[1] = s[1]!;
    simplex[2] = s[2]!;
    fs = fsorted;

    // Convergence: range of f values and simplex size.
    const fRange = Math.abs(fs[2]! - fs[0]!);
    const sizeX = Math.max(
      Math.abs(simplex[1]![0] - simplex[0]![0]),
      Math.abs(simplex[2]![0] - simplex[0]![0])
    );
    const sizeY = Math.max(
      Math.abs(simplex[1]![1] - simplex[0]![1]),
      Math.abs(simplex[2]![1] - simplex[0]![1])
    );
    if (fRange < tol && sizeX < tol && sizeY < tol) break;

    // Centroid of best two.
    const cx = (simplex[0]![0] + simplex[1]![0]) / 2;
    const cy = (simplex[0]![1] + simplex[1]![1]) / 2;

    // Reflection.
    const xr: P = [
      cx + alpha * (cx - simplex[2]![0]),
      cy + alpha * (cy - simplex[2]![1]),
    ];
    const fr = f(xr);
    if (fr < fs[1]! && fr >= fs[0]!) {
      simplex[2] = xr;
      fs[2] = fr;
      continue;
    }
    // Expansion.
    if (fr < fs[0]!) {
      const xe: P = [cx + gamma * (xr[0] - cx), cy + gamma * (xr[1] - cy)];
      const fe = f(xe);
      if (fe < fr) {
        simplex[2] = xe;
        fs[2] = fe;
      } else {
        simplex[2] = xr;
        fs[2] = fr;
      }
      continue;
    }
    // Contraction.
    const xc: P = [
      cx + rho * (simplex[2]![0] - cx),
      cy + rho * (simplex[2]![1] - cy),
    ];
    const fc = f(xc);
    if (fc < fs[2]!) {
      simplex[2] = xc;
      fs[2] = fc;
      continue;
    }
    // Shrink.
    for (let i = 1; i < 3; i++) {
      simplex[i] = [
        simplex[0]![0] + sigma * (simplex[i]![0] - simplex[0]![0]),
        simplex[0]![1] + sigma * (simplex[i]![1] - simplex[0]![1]),
      ];
      fs[i] = f(simplex[i]!);
    }
  }
  return simplex[0]!;
}

// ---------- registry / JSON helpers ----------

export function loadCalibrator(obj: CalibratorJSON | string): Calibrator {
  const parsed: CalibratorJSON = typeof obj === "string" ? JSON.parse(obj) : obj;
  switch (parsed.kind) {
    case "identity":
      return IdentityCalibrator.fromJSON(parsed);
    case "isotonic":
      return IsotonicCalibrator.fromJSON(parsed);
    case "platt":
      return PlattCalibrator.fromJSON(parsed);
    default:
      throw new Error(`Unknown calibrator kind: ${String(parsed.kind)}`);
  }
}

export function saveCalibrator(cal: Calibrator): CalibratorJSON {
  return cal.toJSON();
}

File: infermap/calibration.py
"""Confidence calibration for infermap.

Transforms raw mapping confidences into calibrated probabilities so that
``confidence == p(correct)`` in aggregate. Used post-assignment only — the
calibrator does not influence which mappings the engine picks, only the
scores attached to them. This keeps F1/top-1/MRR bit-identical while moving
ECE.

Calibrators are fitted offline from ``(raw_score, was_correct)`` pairs
collected by running the engine on labeled cases. See
``benchmark/runners/python/infermap_bench/calibrate.py`` for the fitting
CLI.

Serialization uses JSON via ``to_dict``/``from_dict`` with a ``kind``
discriminator — no opaque binary formats (for security and portability).
"""
from __future__ import annotations

import json
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any

import numpy as np


class Calibrator(ABC):
    """Abstract base: fit on labeled pairs, transform raw scores."""

    kind: str = "abstract"

    @abstractmethod
    def fit(self, scores: np.ndarray, correct: np.ndarray) -> None: ...

    @abstractmethod
    def transform(self, scores: np.ndarray) -> np.ndarray: ...

    @abstractmethod
    def to_dict(self) -> dict[str, Any]: ...

    @classmethod
    @abstractmethod
    def from_dict(cls, d: dict[str, Any]) -> "Calibrator": ...


class IdentityCalibrator(Calibrator):
    """Passthrough — useful as a default and in tests."""

    kind = "identity"

    def fit(self, scores: np.ndarray, correct: np.ndarray) -> None:
        return None

    def transform(self, scores: np.ndarray) -> np.ndarray:
        return np.asarray(scores, dtype=float).copy()

    def to_dict(self) -> dict[str, Any]:
        return {"kind": self.kind}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "IdentityCalibrator":
        return cls()


class IsotonicCalibrator(Calibrator):
    """Monotonic non-parametric calibration via pool-adjacent-violators.

    Stores ``x`` (breakpoint scores) and ``y`` (monotonic calibrated values),
    applied at inference via piecewise-linear interpolation
    (``numpy.interp``), clamped to ``[min(x), max(x)]``.
    """

    kind = "isotonic"

    def __init__(self, x: np.ndarray | None = None, y: np.ndarray | None = None):
        self.x = np.asarray(x, dtype=float) if x is not None else np.array([0.0, 1.0])
        self.y = np.asarray(y, dtype=float) if y is not None else np.array([0.0, 1.0])

    def fit(self, scores: np.ndarray, correct: np.ndarray) -> None:
        scores = np.asarray(scores, dtype=float)
        correct = np.asarray(correct, dtype=float)
        if len(scores) == 0:
            return
        # Sort by score ascending (stable for determinism)
        order = np.argsort(scores, kind="stable")
        xs = scores[order]
        ys = correct[order]
        # Pool-adjacent-violators: maintain stacks of (value, weight).
        vals: list[float] = []
        wts: list[float] = []
        for y in ys:
            vals.append(float(y))
            wts.append(1.0)
            while len(vals) >= 2 and vals[-2] > vals[-1]:
                v2 = vals.pop()
                w2 = wts.pop()
                v1 = vals.pop()
                w1 = wts.pop()
                vals.append((v1 * w1 + v2 * w2) / (w1 + w2))
                wts.append(w1 + w2)
        # Expand pooled values back to per-point fitted ys
        fitted = np.empty_like(ys)
        idx = 0
        for v, w in zip(vals, wts):
            k = int(round(w))
            fitted[idx : idx + k] = v
            idx += k
        # Deduplicate by xs: for identical x, average the fitted values
        uniq_x, inv = np.unique(xs, return_inverse=True)
        uniq_y = np.zeros_like(uniq_x)
        counts = np.zeros_like(uniq_x)
        for i, u in enumerate(inv):
            uniq_y[u] += fitted[i]
            counts[u] += 1
        uniq_y = uniq_y / np.maximum(counts, 1)
        # Enforce monotonicity after dedup (defensive).
        uniq_y = np.maximum.accumulate(uniq_y)
        self.x = uniq_x.astype(float)
        self.y = np.clip(uniq_y, 0.0, 1.0).astype(float)

    def transform(self, scores: np.ndarray) -> np.ndarray:
        scores = np.asarray(scores, dtype=float)
        if len(self.x) == 0:
            return scores.copy()
        return np.interp(scores, self.x, self.y, left=self.y[0], right=self.y[-1])

    def to_dict(self) -> dict[str, Any]:
        return {"kind": self.kind, "x": self.x.tolist(), "y": self.y.tolist()}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "IsotonicCalibrator":
        return cls(x=np.array(d["x"], dtype=float), y=np.array(d["y"], dtype=float))


class PlattCalibrator(Calibrator):
    """Sigmoid calibration: ``p = 1 / (1 + exp(-(a*x + b)))``.

    Fit by minimizing binary cross-entropy via ``scipy.optimize.minimize``.
    """

    kind = "platt"

    def __init__(self, a: float = 1.0, b: float = 0.0):
        self.a = float(a)
        self.b = float(b)

    def fit(self, scores: np.ndarray, correct: np.ndarray) -> None:
        from scipy.optimize import minimize

        scores = np.asarray(scores, dtype=float)
        correct = np.asarray(correct, dtype=float)
        if len(scores) == 0:
            return

        def nll(params: np.ndarray) -> float:
            a, b = params
            z = a * scores + b
            logp = -np.logaddexp(0.0, -z)  # log sigmoid(z)
            log1mp = -np.logaddexp(0.0, z)  # log (1-sigmoid(z))
            return -float(np.sum(correct * logp + (1.0 - correct) * log1mp))

        res = minimize(nll, x0=np.array([1.0, 0.0]), method="Nelder-Mead")
        self.a = float(res.x[0])
        self.b = float(res.x[1])

    def transform(self, scores: np.ndarray) -> np.ndarray:
        scores = np.asarray(scores, dtype=float)
        z = self.a * scores + self.b
        return np.where(z >= 0, 1.0 / (1.0 + np.exp(-z)), np.exp(z) / (1.0 + np.exp(z)))

    def to_dict(self) -> dict[str, Any]:
        return {"kind": self.kind, "a": self.a, "b": self.b}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "PlattCalibrator":
        return cls(a=float(d["a"]), b=float(d["b"]))


_REGISTRY: dict[str, type[Calibrator]] = {
    "identity": IdentityCalibrator,
    "isotonic": IsotonicCalibrator,
    "platt": PlattCalibrator,
}


def calibrator_from_dict(d: dict[str, Any]) -> Calibrator:
    kind = d.get("kind")
    if kind not in _REGISTRY:
        raise ValueError(f"Unknown calibrator kind: {kind!r}")
    return _REGISTRY[kind].from_dict(d)


def save_calibrator(cal: Calibrator, path: str | Path) -> None:
    Path(path).write_text(json.dumps(cal.to_dict(), indent=2), encoding="utf-8")


def load_calibrator(path: str | Path) -> Calibrator:
    d = json.loads(Path(path).read_text(encoding="utf-8"))
    return calibrator_from_dict(d)


# TODO(ts-port): TypeScript parity. Port IdentityCalibrator, IsotonicCalibrator,
# and PlattCalibrator to infermap-js with the same JSON schema
# ({kind, x, y} / {kind, a, b}) so Python-fitted calibrators can be loaded by
# the JS engine. PAV is ~40 LOC with no heavy deps; Platt needs a small MLE
# (hand-rolled Nelder-Mead or gradient descent) since scipy isn't available in
# JS. Tracked as a follow-up.
